opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py

```diff
@@ -349,6 +362,334 @@ def collect(
+        self._start_time_unix_nano = start_time_unix_nano
+
+    def aggregate(self, measurement: Measurement) -> None:
+        # pylint: disable=too-many-branches,too-many-statements
+
+        value = measurement.value
+
+        # 0. Set the following attributes:
+        # _min
+        # _max
+        # _count
+        # _zero_count
+        # _sum
+        if value < self._min:
+            self._min = value
+
+        if value > self._max:
+            self._max = value
+
+        self._count += 1
+
+        if value == 0:
+            self._zero_count += 1
+            return
+
+        self._sum += value
+
+        # 1. Use the positive buckets for positive values and the negative
+        # buckets for negative values.
+        if value > 0:
+            buckets = self._positive
+
+        else:
+            # Both exponential and logarithm mappings use only positive values
+            # so the absolute value is used here.
+            value = -value
+            buckets = self._negative
+
+        # 2. Compute the index for the value at the current scale.
+        index = self._mapping.map_to_index(value)
+
+        # 3. Determine if a change of scale is needed.
+        is_rescaling_needed = False
+
+        if len(buckets) == 0:
+            buckets._index_start = index
+            buckets._index_end = index
+            buckets._index_base = index
+
+        elif (
+            index < buckets._index_start
+            and (buckets._index_end - index) >= self._max_size
+        ):
+            is_rescaling_needed = True
+            low = index
+            high = buckets._index_end
+
+        elif (
+            index > buckets._index_end
+            and (index - buckets._index_start) >= self._max_size
+        ):
+            is_rescaling_needed = True
+            low = buckets._index_start
+            high = index
+
+        # 4. Rescale the mapping if needed.
+        if is_rescaling_needed:
+
+            change = 0
+
+            while high - low >= self._max_size:
+                high = high >> 1
+                low = low >> 1
+
+                change += 1
+
+            if change < 0:
+                raise Exception(f"Impossible change of scale: {change}")
+
+            new_scale = self._mapping.scale - change
+
+            self._positive.downscale(change)
+            self._negative.downscale(change)
+
+            if new_scale <= 0:
+                mapping = ExponentMapping(new_scale)
+            else:
+                mapping = LogarithmMapping(new_scale)
+
+            self._mapping = mapping
+
+            index = self._mapping.map_to_index(value)
+
+            if (
+                index < buckets._index_start
+                and (buckets._index_end - index) >= self._max_size
+            ) or (
+                index > buckets._index_end
+                and (index - buckets._index_start) >= self._max_size
+            ):
+                raise Exception("Downscaling logic error")
```

Comment:
Can you make a more descriptive message and ideally specific exception type?

Should we be throwing here at all?

Reply:
Right, I removed it.